### ISSUE TITLE
[v8] Port React-Redux to TypeScript

### ISSUE THREAD

**Issue by markerikson** [reduxjs/react-redux#1737]
React-Redux is currently written in plain JS, and our types have been maintained by the community over in DefinitelyTyped at https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react-redux/ .

Whenever we get around to writing a React-Redux version 8, we intend to port the library to be written in TypeScript directly.

While we don't yet have plans to work on a new version, it's likely that might happen in the next few months because of React 18 and the Concurrent Rendering changes in that release.

It would be helpful if we could start the work of porting React-Redux to TS now, even if it's just a draft on a branch, so that we have that complex work done ahead of time and can then look at any compat changes for React 18 separately without trying to do both at once.

We maintainers probably don't have the time to tackle this ourselves right now, but I'd love to see the community work on that port so we get it ready.

**Comment by chuckdries** [comment]
I'm down to help! If no one else has done so by the time I get off work today, I'll spin up a branch and at least configure typescript so we can start working.

**Comment by markerikson** [comment]
Tell you what. In the interest of keeping collaboration centralized, I'm going to go ahead and create an integration branch here in the main repo.  That way we can have PRs and code progress in one place.

**Comment by markerikson** [comment]
Okay, I just created a new https://github.com/reduxjs/react-redux/tree/typescript-port branch .  Let's use that as the target for all PRs for this effort.

A couple points to consider:

- The goal for now is to exactly match current React-Redux v7.2.x behavior in both logic and TS.  In other words, we're not trying to add any new features or change any TS types, _for now_.  I want an exact match of current functionality as a baseline. _Then_ we can use that as a starting point for any future logic or types changes.  That does mean porting over all the ridiculously complicated helper types and overloads that are currently in the React-Redux typedefs.
- We might want to consider pulling over the ESBuild-powered build toolchain that we recently set up for Redux Toolkit.  It's slightly less critical here because we aren't going to have multiple entry points the way we now have in RTK, but it's something that we know works and is probably less complicated than using TSDX or similar.  See [the `scripts` folder in the RTK repo](https://github.com/reduxjs/redux-toolkit/tree/583823474d041773229a8e78b2ceba37f7616851/scripts) for the current source of the build toolchain.

**Comment by vinay-deshmukh** [comment]
@markerikson 

Is there any plan / structure on how contributors should start converting .js to .ts? 
For example, first the `hooks` folder, then the `connect` folder.

So that, we don't get two people working on the same set of files at the same time.

Perhaps something like "sub-issues" to this issue to keep track of which files are being worked on.

**Comment by markerikson** [comment]
What you see here is all the discussion so far :)

I can help provide some direction and coordination a bit, but I don't have time to be seriously active with this effort for now.  So, this needs to be a collaborative thing.  That includes filing issues or anything else that needs to be done to help track the work.

**Comment by markerikson** [comment]
Okay, we've got some basic TS build integration in place, and the first couple files (context and Provider) ported over.

I suspect some of the pieces of `connect` could be done one file at a time, particularly the selector/`map*` related files.

**Comment by OsidAbu-alrub** [comment]
Hello, I'd love to help

**Comment by tony-go** [comment]
Hi Guys 👋 

Very nice initiative 🙌 

I'd like to start something on `connectAdvanced.js` and deps as`useIsomorphicLayoutEffect` (`Subscription` seems to have been migrated already ).

Let me know If it's good. 

@markerikson If you think I could start on something "better". Let me know ^^

**Comment by markerikson** [comment]
I just merged #1743 , which converted several of the smaller files to TS.

If we look at the https://github.com/reduxjs/react-redux/tree/typescript-port branch , here's what's left file-wise atm:

` `/src`:
  - the entry point files
- `/src/components`
  - `connectAdvanced.js`
- `/src/connect`
  - all files (`mapStateToProps.js`, etc)
- `/src/hooks`:
  - `useSelector.js`
- `/src/utils`
  - the `batchedUpdates` files
  
There's also all the test files that will need to be converted over, and I don't know if we have the test setup itself configured to build / run TS files yet.  We also will probably need to copy-paste over additional types and typedefs from the current `@types/react-redux` package.

I think we ought to save the three biggest pieces of the actual source for last: `useSelector.js`, `connect.js`, and `connectAdvanced.js`.

So, I'd suggest folks should claim a couple of the other files in `/connect` and work on those first.

**Comment by tony-go** [comment]
Well @markerikson ^^

I'll start something on `mapStateToProps`. 

Ok for y'all ?

**Comment by markerikson** [comment]
Yep, go for it.

**Comment by myNameIsDu** [comment]
I want to help. Is there anything I can do

**Comment by markerikson** [comment]
@myNameIsDu : no one's claimed it yet, but I think it might be good if I try doing that one myself. Should have some time to try it this weekend.

**Comment by myNameIsDu** [comment]
> @myNameIsDu : no one's claimed it yet, but I think it might be good if I try doing that one myself. Should have some time to try it this weekend.

Well, if there's anything I can do, please let me know. I'd be happy to take part in it

**Comment by tony-go** [comment]
Hi there 👋 

I can follow up with [`mapDispatchToProps`](https://github.com/reduxjs/react-redux/blob/typescript-port/src/connect/mapDispatchToProps.js) WDYT ?

**Comment by markerikson** [comment]
@tony-go yep, go right ahead!

**Comment by CoolRaja001** [comment]
How I make website like [LyricsGoal](https://www.lyricsgoal.com/)

**Comment by markerikson** [comment]
As a heads-up: we recently converted the RTK repo to Yarn v2, and I'm working to convert the React-Redux repo to Yarn v2 based on this `typescript-port` branch as well.

If anyone else is planning to do PRs for this, be aware that this change is coming (hopefully shortly).

**Comment by markerikson** [comment]
Yarn v2 switch done via #1751 .  Please update your clones/forks as needed.

**Comment by markerikson** [comment]
FYI, I'm starting to tackle `connectAdvanced.js`, which also strongly overlaps with `connect.js` and `selectorFactory.js`.  @tony-go is doing `mapDispatchToProps.js`.

I _think_ this is the last of the meaningful files that need to be converted. However:

- The entry point files need to be converted, and need to export all relevant types
- We still need to get our test files converted to TS
- We don't have the build setup outputting TS typedefs

So, still plenty left to do here.

**Comment by myNameIsDu** [comment]
@markerikson  Can I complete the entry file conversion and export all types of tasks？

**Comment by markerikson** [comment]
@myNameIsDu yeah, go for it! I just merged the `connect` types PR.

**Comment by markerikson** [comment]
## TypeScript Port Status Update

I'm happy to report that **all React-Redux source files have now been converted to TypeScript!**

This does _not_ mean that all the types are _correct_, of course :)  I suspect we have a lot of work to do to try to match the existing typedefs in terms of exports and behavior.  (I'll be very surprised if `connect` is carrying the right input types through the entire process atm.)

Turns out we do actually have TS typedefs for our source files being generated - I just hadn't looked in the right spot. The initial TS build conversion PR apparently put them into `/lib` in our output, but I've switched that to be `/es` instead.

### Next Steps

We need to evaluate the current types in the code, and the types we're outputting, and try to get them to match the current `@types/react-redux` behavior as much as possible in terms of what types are publicly visible in the output files and how those types behave.  

I'm not quite sure what the best approach is for this.  It would probably help if we had some type tests that specifically target `react-redux@7.2.4` + `@types/react-redux`, and then can try writing equivalent type tests that use the output of this converted codebase.  We can also try setting up `api-extractor`, and perhaps we can figure out a way to get it to generate an output example for `@types/react-redux` that we can compare against.

We also still need to convert the rest of our test files to TS as well.

Once we accomplish those tasks, we can create a `7.x` branch off `master`, merge #1739 into `master`, and then do any further 8.x development work there.

**Comment by myNameIsDu** [comment]
I'll do the test file conversion. I have some time on the weekend :smile:

**Comment by StevenLangbroek** [comment]
Hey @myNameIsDu! I started in #1765, do you want to work on this together?

**Comment by myNameIsDu** [comment]
> Hey @myNameIsDu! I started in #1765, do you want to work on this together?

Thanks for inviting me. Let's get it done together

**Comment by markerikson** [comment]
I just did a bunch of reworking of the `connect` types to better match what's in the existing `@types/react-redux` typedefs.  I also added a typetests file based on what was in DT.

@StevenLangbroek , @myNameIsDu : I think there's a good chance this _could_ fix whatever type issues you were seeing while converting tests. Try merging/rebasing and let me know what happens.

**Comment by myNameIsDu** [comment]
@StevenLangbroek How should I cooperate with you？:smiley:

**Comment by myNameIsDu** [comment]
OK, I will start by fixing the file under test=>hooks. If you want to get involved, please be careful not to clash

**Comment by myNameIsDu** [comment]
I will fixing the file under test=>components. If you want to get involved, please be careful not to clash :smiley:

**Comment by StevenLangbroek** [comment]
Hey @myNameIsDu! So regarding how to collaborate, I think (given the fact that Github doesn't support Stacked PRs) we have 2 options, either is fine for me:

1. You work in your own fork, submit PRs against my fork/branch (since it has the basic setup there)
2. I give you maintainer access to my fork, we can work in the same branch if we go folder by folder / task by task

I think 2 is easiest, if you'd rather work some other way I'd love to hear it.

**Comment by myNameIsDu** [comment]
> Hey @myNameIsDu! So regarding how to collaborate, I think (given the fact that Github doesn't support Stacked PRs) we have 2 options, either is fine for me:
> 
> 1. You work in your own fork, submit PRs against my fork/branch (since it has the basic setup there)
> 2. I give you maintainer access to my fork, we can work in the same branch if we go folder by folder / task by task
> 
> I think 2 is easiest, if you'd rather work some other way I'd love to hear it.

I don't think it's that troublesome, We just need to separate our files so that there is no conflict :smirk:

However, I also think the second of these two methods easier

**Comment by myNameIsDu** [comment]
@markerikson   If want to delete ConnectAdvanced, will the test case be rewritten? If so, I will not continue to convert it

**Comment by markerikson** [comment]
Yeah, you can see that my PR in #1771 deletes that test case file. So, ignore that - it'll go away when #1771 gets merged.

**Comment by myNameIsDu** [comment]
@markerikson What do you need me to do next? Because you are changing to connect, I did not change connect.test

**Comment by markerikson** [comment]
@myNameIsDu The `connect` tests file _should_ stay basically the same - that's intentional.  I think I did delete one or two tests in that PR, but that's it.

It's just that `connectAdvanced` is going away.

**Comment by myNameIsDu** [comment]
> @myNameIsDu The `connect` tests file _should_ stay basically the same - that's intentional. I think I did delete one or two tests in that PR, but that's it.
> 
> It's just that `connectAdvanced` is going away.
> 
> I'd say go ahead and convert `connect.spec.ts` to TS right now, and I'll take care of doing the merge with that other PR myself later.

Ok, I'm going to convert connect.spec.js, which is exciting

**Comment by myNameIsDu** [comment]
> @myNameIsDu The `connect` tests file _should_ stay basically the same - that's intentional. I think I did delete one or two tests in that PR, but that's it.
> 
> It's just that `connectAdvanced` is going away.
> 
> I'd say go ahead and convert `connect.spec.ts` to TS right now, and I'll take care of doing the merge with that other PR myself later.

Ok, I'm going to convert connect.spec.js, which is exciting  :smirk:

**Comment by markerikson** [comment]
I've just merged the TS integration PR at #1739 into `master`!

I created a `7.x` branch off of master right before I merged, and we'll use that for any future maintenance fixes on v7.

We still have some test files left to convert, but we can continue that work against `master`.  I'll leave the issue open just until we get the rest of the files converted, then close it.

Thank you SO much to everyone who's contributed to this effort!
